store-implementation/federated-store/src/main/java/uk/gov/gchq/gaffer/federatedstore/FederatedGraphStorage.java

```diff
@@ -0,0 +1,193 @@
+            throw new OverwritingException((String.format(USER_IS_ATTEMPTING_TO_OVERWRITE_A_GRAPH_WITHIN_FEDERATED_STORE_GRAPH_ID_S, graph.getGraphId())));
+        } else if (null == access) {
+            throw new RuntimeException(CAN_NOT_PUT_GRAPH_INTO_STORAGE_WITHOUT_A_FEDERATED_ACCESS_KEY);
+        }
+
+        // Check the schema can be merged before adding the graph.
+        getMergedSchema(Optional.of(graph));
+
+        Set<Graph> existingGraphs = storage.get(access);
+        if (null == existingGraphs) {
+            existingGraphs = Sets.newHashSet(graph);
+            storage.put(access, existingGraphs);
+        } else {
+            existingGraphs.add(graph);
+        }
+
+        mergedSchema = getMergedSchema(Optional.empty());
+        mergedTraits = getMergedTraits(Optional.empty());
+    }
+
+    public void remove(final String graphId) {
+        for (final Entry<FederatedAccess, Set<Graph>> entry : storage.entrySet()) {
+            final Set<Graph> graphs = entry.getValue();
+            if (null != graphs) {
+                for (final Graph graph : graphs) {
+                    if (graph.getGraphId().equals(graphId)) {
+                        graphs.remove(graph);
+                    }
+                }
+            }
+        }
+        mergedSchema = getMergedSchema(Optional.empty());
+        mergedTraits = getMergedTraits(Optional.empty());
+    }
+
+    public Schema getMergedSchema() {
+        return mergedSchema.clone();
+    }
+
+    private boolean exists(final String graphId) {
+        boolean exists = false;
+        outer:
+        for (final Set<Graph> graphs : storage.values()) {
+            for (final Graph graph : graphs) {
+                if (graph.getGraphId().equals(graphId)) {
+                    exists = true;
+                    break outer;
+                }
+            }
+        }
+        return exists;
+    }
+
+    public Set<StoreTrait> getTraits() {
+        return Collections.unmodifiableSet(mergedTraits);
+    }
+
+    private Schema getMergedSchema(final Optional<Graph> newGraph) {
+        Builder schemaBuilder = getMergedSchemaBuilder(Optional.empty());
+
+        if (newGraph.isPresent()) {
+            schemaBuilder.merge(newGraph.get().getSchema());
+//        An exception would be thrown here if something was wrong merging the schema.
+        }
+
+        return schemaBuilder.build();
+    }
+
+    private Builder getMergedSchemaBuilder(final Optional<User> user) {
+        Builder schemaBuilder = new Builder();
+        storage.entrySet()
+                .stream()
+                .filter(filterByOptionalUser(user))
+                .flatMap(entry -> entry.getValue().stream()).forEach(graph -> schemaBuilder.merge(graph.getSchema()));
+        return schemaBuilder;
+    }
+
+    public Set<StoreTrait> getTraits(final User user) {
+        return getMergedTraits(Optional.of(user));
+    }
+
+    private Set<StoreTrait> getMergedTraits(final Optional<User> user) {
```

Comment:
I think some of this logic could be simplified and some of the methods could be removed or some of the arguments could be removed.